subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1281 @@
+#define FLAG_ICMSG_BOUNDED BIT(31)
+
+/** Registered endpoints count mask in flags. */
+#define FLAG_EPT_COUNT_MASK 0xFFFF
+
+enum msg_type {
+	MSG_DATA = 0,		/* Data message. */
+	MSG_RELEASE_DATA,	/* Release data buffer message. */
+	MSG_BOUND,		/* Endpoint bounding message. */
+	MSG_RELEASE_BOUND,	/* Release endpoint bound message. */
+};
+
+enum ept_bounding_state {
+	EPT_UNCONFIGURED = 0,	/* Endpoint in not configured (initial state). */
+	EPT_CONFIGURED,		/* Endpoint is configured, waiting for work queue to
+				 * start bounding process.
+				 */
+	EPT_BOUNDING,		/* Only on initiator. Bound message was send,
+				 * but bound callback was not called yet, because
+				 * we are waiting for any incoming messages.
+				 */
+	EPT_READY,		/* Bounding is done. Bound callback was called. */
+};
+
+struct channel_config {
+	uint8_t *blocks_ptr;	/* Address where the blocks start. */
+	size_t block_size;	/* Size of one block. */
+	size_t block_count;	/* Number of blocks. */
+};
+
+struct icmsg_with_buf_config {
+	struct icmsg_config_t icmsg_config;	/* Configuration of the ICMsg. */
+	struct channel_config rx;		/* RX channel config. */
+	struct channel_config tx;		/* TX channel config. */
+	sys_bitarray_t *tx_usage_bitmap;	/* Bit is set when TX block is in use */
+	sys_bitarray_t *rx_hold_bitmap;		/* Bit is set, if the buffer starting at
+						 * this block should be kept after exit
+						 * from receive handler.
+						 */
+};
+
+struct ept_data {
+	const struct ipc_ept_cfg *cfg;	/* Endpoint configuration. */
+	atomic_t state;			/* Bounding state. */
+	uint8_t addr;			/* Endpoint address. */
+};
+
+struct backend_data {
+	const struct icmsg_with_buf_config *conf; /* Backend instance config. */
+	struct icmsg_data_t icmsg_data;	/* ICMsg data. */
+	struct k_mutex mutex;		/* Mutex to protect: ICMsg send call and
+					 * waiting_bound field.
+					 */
+	struct k_work ep_bound_work;	/* Work item for bounding processing. */
+	struct k_sem block_wait_sem;	/* Semaphore for waiting for free blocks. */
+	struct ept_data ept[NUM_EPT];	/* Array of registered endpoints. */
+	uint8_t ept_map[NUM_EPT];	/* Array that maps endpoint address to index. */
+	uint16_t waiting_bound[NUM_EPT];/* The bound messages waiting to be registered. */
+	atomic_t flags;			/* Flags on higher bits, number of registered
+					 * endpoints on lower.
+					 */
+	bool is_initiator;		/* This side has an initiator role. */
+};
+
+struct block_header {
+	volatile size_t size;	/* Size of the data field. It must be volatile, because
+				 * when this value is read and validated for security
+				 * reasons, compiler cannot generate code that reads
+				 * it again after validation.
+				 */
+	uint8_t data[];		/* Buffer data. */
+};
+
+struct icmsg_message {
+	uint8_t msg_type;	/* Message type. */
+	uint8_t ept_addr;	/* Endpoint address or zero for MSG_RELEASE_DATA. */
+	uint8_t block_index;	/* Block index to send or release. */
```

Comment:
I moved block count to the DT instead of kconfig, because I assume different instances talks with different CPUs. In "kconfig" case, all CPUs would need the same block count (or size). I think that we need more flexibility, since there is no almost additional cost putting it in DT.

Regarding `block count` vs `block size`, I don't have very strong arguments for any of the those approaches. They both have advantages and disadvantages. I simply choose the approach that was more natural for me.

I think that runtime block size exchange is not a good solution. I see no benefits over DT option except user provides block count just on one side. The problem is that a lot of code will be added to handle that. Currently most of block count/size/address calculations are done in build time.